atsign-foundation/at_c
# feat: notify encryption + functional test

closes #216 

**- What I did**
- notify encryption
- atclient_notify_create_params function
- added `shouldencrypt` and `sharedenckeybase64` in `notify_params` struct
- Misc docs changes here and there

Stuff that works and don't:
- examples/desktop/at_talk - compiles but untested
- examples/desktop/events/notify.c - compiles and tested
- examples/desktop/events/monitor.c - compiles but untested

Changes from @realvarx and @XavierChanth :
- new `monitor.c/h`
- refactored `notification.c/h` to `notify.c/h`
- atkey_from_string enhancement (better namespace checking I think?)
- atclient_atkey_metadata_from_cjson_node enhancement (handles "null" values)
- tools/dev.sh

**- How I did it**

**- How to verify it**
- See functional tests
- `test_atclient_notify.c`

```c
static int test_1_notify(atclient *atclient, char *notification_id) {
  int ret = 1;

  atclient_notify_params params;
  atclient_notify_params_init(&params);

  atclient_atkey atkey;
  atclient_atkey_init(&atkey);

  if ((ret = atclient_atkey_create_sharedkey(&atkey, ATKEY_KEY, strlen(ATKEY_KEY), ATKEY_SHAREDBY,
                                             strlen(ATKEY_SHAREDBY), ATKEY_SHAREDWITH, strlen(ATKEY_SHAREDWITH),
                                             ATKEY_NAMESPACE, strlen(ATKEY_NAMESPACE))) != 0) {
    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to create atkey: %d\n", ret);
    goto exit;
  }

  atclient_notify_params_create(&params, ATNOTIFICATION_OPERATION, &atkey, ATKEY_VALUE);

  if ((ret = atclient_notify(atclient, &params, NULL)) != 0) {
    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to notify: %d\n", ret);
    goto exit;
  }

  goto exit;

exit: {
  atclient_notify_params_free(&params);
  atclient_atkey_free(&atkey);
  return ret;
}
}
```

**- Description for the changelog**
feat: notify encryption + functional test
